### PR TITLE
Making Requests Stable

### PR DESCRIPTION
**Pull request by smusali**
Adding `retries` and `backoff_factor` for establishing connection with `LogDNA Ingesters` should improve the stability of the code library since recently the following issue was reported:
```
Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f6feb2b5c18>: Failed to establish a new connection: [Errno -2] Name or service not known',)
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 **/*.pyc
+**/*cache*
```
Comment (dchai76) [inline .gitignore]: Add a newline please

File: logdna/logdna.py [hunk 2]
```diff
@@ -1,72 +1,83 @@
...
 import time
 import json
-from .configs import defaults
 import logging
-import requests
-import threading
 import socket
-from threading import Timer
+import threading
+import requests
```
Comment (dchai76) [inline logdna/logdna.py]: Pythoners (but actually all coders) tend to alphabetize their imports. I wouldn't mind you doing that here.

File: logdna/logdna.py [hunk 3]
```diff
@@ -1,72 +1,83 @@
...
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+from .configs import defaults
+from .utils import sanitize_meta
+from .utils import get_ip
+from .utils import get_option
```
Comment (dchai76) [inline logdna/logdna.py]: Format-wise, I'm more used to `from .utils import get_ip, get_option, sanitize_meta`

File: logdna/utils.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        # doesn't even have to be reachable
+        s.connect(('10.255.255.255', 1))
+        ip = s.getsockname()[0]
+    except:
+        ip = '127.0.0.1'
+    finally:
+        s.close()
+    return ip
+
+def get_option(options, key, default=None):
```
Comment (dchai76) [inline logdna/utils.py]: You don't need this, this functionality is built into Python. Use `options.get(key, default)`.

**Comment by respectus** [comment]
These improvements look good, let's get them tested.

File: logdna/logdna.py (hunk 2 above)
Comment (smusali) [reply logdna/logdna.py]: `pylint` was complaining for the order - after fixing the order, didn't care about having alphabetical order that much.

**Comment by smusali** [comment]
@dchai76, updated! Can you do second round?

File: logdna/logdna.py [hunk 5]
```diff
@@ -1,72 +1,81 @@
...
+        self.exception_flag = False
+        self.buf_byte_length = 0
         self.flusher = None
+        self.max_length = defaults['MAX_LINE_LENGTH']
+        self.connection_retries = 5
+        self.retry_backoff_factor = 0.5
+
+        self.hostname = options.get('hostname', socket.gethostname())
+        self.ip = options.get('ip', get_ip())
+        self.mac = options.get('mac', None)
```
Comment (dchai76) [inline logdna/logdna.py]: `None` is the default for `get()` so you don't need that. But it's fine to keep.

**Comment by danmaas** [comment]
Not sure if you intend to keep supporting Python 2, but the following update to `setup.py` breaks Python 2.x because it doesn't have the "encoding" option for open():

```
with open(path.join(this_directory, 'README.md'), encoding='utf-8') as f:
```

error is:
```
TypeError: 'encoding' is an invalid keyword argument for this function
```

For info and work-around see [https://stackoverflow.com/questions/10971033/backporting-python-3-openencoding-utf-8-to-python-2](https://stackoverflow.com/questions/10971033/backporting-python-3-openencoding-utf-8-to-python-2)

**Comment by smusali** [comment]
Thanks for bringing this to our attention, @danmaas! Fixed in [`v1.4.2`](https://pypi.org/project/logdna/)

**Pull request merged**